README.md
# [@PyFunceble](https://pyfunceble.github.io) on miniconda

[![Codacy Badge](https://api.codacy.com/project/badge/Grade/7e9d0f339cd046c4ab8b162e6805f182)](https://app.codacy.com/gh/PyFunceble-Templates/pyfunceble-miniconda?utm_source=github.com&utm_medium=referral&utm_content=PyFunceble-Templates/pyfunceble-miniconda&utm_campaign=Badge_Grade_Settings)

This script consist of two parts.

  1. `setup_miniconda.sh`
  2. `pyfunceble.sh`

## Setup

If everything works as expected, you should never run `setup_miniconda.sh`
more than once. This is because we will be using the .environment.yaml
to keep your version up to date.

## Running PyFunceble

Simple, feat the scripts with answers or it staves and die... :skull:

For testing purpose you can use the default @PyFunceble test file from
<https://raw.githubusercontent.com/PyFunceble/ci_test/master/test.list>

The command would then be

```shell
$ ./pyfunceble.sh "https://raw.githubusercontent.com/PyFunceble/ci_test/master/test.list"
```

## Read more

You can read more about environment variables at
<https://pyfunceble.readthedocs.io/en/dev/usage/index.html#from-a-terminal>

## Test string

To alter the test options you should be reading more about this at
<https://pyfunceble.readthedocs.io/en/dev/usage/index.html>

## Why this script

See <https://github.com/funilrys/PyFunceble/issues/39> by
[@mitchellkrogza](https://github.com/mitchellkrogza)
and @funilrys and since I'm running on a Ubuntu
which curses some troubles
as [@mitchellkrogza](https://github.com/mitchellkrogza) describes it:

> Distributions like Ubuntu are especially troublesome with Python issues
> which are easily solved by just running Python in Conda environments.


CHANGELOG
1.0.2 Fixed a old hanging rename issue

1.0.1 Changeset to fit PyFunceble 4.0.1
      - Replacement of bash code lines for installing into conda via
        `.environment.${pyfunceblePackageName}.yaml` file.
      - Added this CHANGELOG
      - Updated `README.md`, A few typos and added section to the "Setup"
        about the future update/upgrade should be happening through the
        standard script, as it is related to the
        `.environment.${pyfunceblePackageName}.yaml` file.
      - Support for arm7 have been dropped
      - Moved `pyfunceble4.sh` to `pyfunceble.sh`
      - Added copying the `.PyFunceble.overwrite.yaml`
      - Updated `requirements.txt`
      - added test of default conda install dir from the install script
        to the `pyfunceble.sh`
      - Changed the selection of PyFunceble version in `pyfunceble.sh`
        to a numbered choice rather than spelling.
      - Set the installer script to delete `miniconda.sh` after
        installation
      - replaced the rm command with find -delete


pyfunceble.sh
#!/usr/bin/env bash
# All initial credits to @mitchellkrogza and @funilrys
# Modified by @spirillen
# License: https://mypdns.org/mypdns/support/-/wikis/License
# Issues: https://mypdns.org/pyfunceble/pyfunceble-templates/pyfunceble-miniconda/-/issues
# Project: https://mypdns.org/pyfunceble/pyfunceble-templates/pyfunceble-miniconda
# -------------------------------
# Setup Conda Python Environments
# -------------------------------

# Stop on any error
set -e #-x

SCRIPT_DIR=$(DIRNAME=$(dirname "$0"); cd "$DIRNAME"; pwd)
# THIS_FILE=$(basename "$0")

pkgs="tree"
if ! dpkg -s $pkgs >/dev/null 2>&1; then
  sudo apt-get install $pkgs
fi

# Run this script by appending test-file to the script name in the shell prompt
# E.g. ./pyfunceble.sh "/full/path/to/file"

# just a bit of fun
echo ""
echo -e "\tI'm your hungry cheese missing a piece..."
echo -e "\tAre you ready to go hunting the red pills"
echo -e "\tso we can hunt down evil ghosts"
echo ""
echo -e "\tAre you ready to start?"
echo ""
echo -e "\tloading kacman..."
echo ""

# Set conda install dir
condaInstallDir="${HOME}/miniconda"

if [ ! -d "$condaInstallDir" ]
then
    echo "Please install miniconda"
    echo "run './setup_miniconda.sh' from $SCRIPT_DIR"
    exit 10
fi

if [[ -z "${1}" ]]
then
    echo ""
    echo -e "\n\tYou have been swallowing a blue pill..."
    echo -e "\tThe ghosts caught you :skull:"
    echo -e "\tPlease show me the route to the ghost's"
    echo -e "\tYou want me to chew through\n\t.%s /blue/pills/is/dead/ghosts\n\n" "${0}"
    exit 1
fi

# Select Python version for test env
# read -erp "Enter the Python version you like to use within Conda: " \
#     -i "3.10" PV

# Change the output directory to suite your needs
read -erp "Enter output directory for test results: " \
    -i "/tmp/pyfunceble/$(date +'%H%M')" outputDir

# Clean output dir if exist for a clean test environment
if [[ -d "${outputDir}" ]]
then
    # rm -fr "${outputDir}*"
    find "${outputDir}" -mindepth 1 -delete
fi

# Set your desired pyfunceble version
# We set pyfunceble-dev as default to avoid typos.

# We should make this a numbered optional list
echo ""
echo "You can choose between the following versions of PyFunceble"
echo "pyfunceble latest Stable " "Option = 1"
echo "pyfunceble latest released develop version " "Option = 2"
echo ""
echo ""

read -erp "Which version of PyFunceble would you like to use?: " \
    -i "2" PYF_OPTION

if [ "$PYF_OPTION" == "1" ]
then
    PYFUNCEBLEPACKAGENAME="pyfunceble"
else
    PYFUNCEBLEPACKAGENAME="pyfunceble-dev"
fi

read -erp "Should we delete the Conda ENV after this job completed?: " \
    -i "Y" CCLEAN

# Set your test string.
echo ""
echo "IMPORTANT: the -f argument is preset as last argument"
echo ""

# Bug #3 test string
# read -erp "Enter any custom test string: " \
#     -i "--hierarchical -ex -a --dns 192.168.1.6 -w 40 -dbr 0 --whois-lookup 
#         --database-type mariadb  --http --netinfo-lookup --push-collection" \
#         -a PYFUNCEBLEARGS

read -erp "Enter any custom test string: " -i "--dns 192.168.1.6 -w 40" \
    -a PYFUNCEBLEARGS

echo ""
# We should change the default ENV dir to match the PyF versions conda dir
# shellcheck disable=SC2034  # Unused variables left for readability

while true
do
read -erp "Would you like to use your default pyfunceble environment
  ${condaInstallDir}/envs/${PYFUNCEBLEPACKAGENAME}?: [Y/n] " -i "Y" PYFUNCEBLE_ENV

case $PYFUNCEBLE_ENV in
    [yY][eE][sS]|[yY])
 useEnvPath="yes"
 break
 ;;
    [nN][oO]|[nN])
 useEnvPath=""
 break
    ;;
    *)
 echo "Invalid input..."
 ;;
 esac
done

# Get the conda CLI.
source "${condaInstallDir}/etc/profile.d/conda.sh"

hash conda

# conda config --set changeps1 true

# First Update Conda
conda update -q conda

# Activate your environment
# According to the https://docs.conda.io/projects/conda/en/latest/_downloads/843d9e0198f2a193a3484886fa28163c/conda-cheatsheet.pdf
# We shall replace source with conda activate vs source
# conda env update -f "$SCRIPT_DIR/.environment.yaml"
conda env update -f "$SCRIPT_DIR/.environment.${PYFUNCEBLEPACKAGENAME}.yaml" --prune -q

# # Setup Conda Env if it does not exist, else update
# if [ ! -d "${condaInstallDir}/envs/${PYFUNCEBLEPACKAGENAME}" ]
# then
#     conda create -n "${PYFUNCEBLEPACKAGENAME}" python="${PV}" -yqq
# else
#     conda activate "${PYFUNCEBLEPACKAGENAME}"

#     # Set python version to use
#     conda install python="${PV}"
# fi

# Make sure output dir is present
mkdir -p "${outputDir}"

# Upgrade the environment
# pip install -I -U -q pip wheel Pygments>=2.0
# pip uninstall -yq PyFunceble-dev #"${PYFUNCEBLEPACKAGENAME}"
# pip install --no-cache-dir --upgrade -q --pre pyfunceble-dev>=4.0.0b62 'alabaster<0.8,>=0.7'
# pip install --no-cache-dir --upgrade -I -q 'git+https://mypdns.org/funilrys/PyFunceble.git@dev'
# pip install --no-cache-dir --upgrade -I -q 'git+https://mypdns.org/pyfunceble/PyFunceble.git@dev'
# pip install --no-cache-dir --upgrade -I -q -r "$SCRIPT_DIR/requirements.txt"

# if [ "${PYFUNCEBLEPACKAGENAME}" == 'pyfunceble' ]
# then
#     pip install --no-cache-dir --upgrade -I -q 'git+https://github.com/Ultimate-Hosts-Blacklist/whitelist.git@script'
# 	pyfunceble --version
# 	uhb-whitelist --version
# 	pip list
# else
#     pip install --no-cache-dir --upgrade -I -q 'git+https://github.com/Ultimate-Hosts-Blacklist/whitelist.git@script-dev'
# 	pyfunceble --version
# 	uhb-whitelist --version
# 	pip list
# fi


hash pyfunceble

# print pyfunceble version
pyfunceble --version

# Tell the script to install/update the configuration file automatically.
export PYFUNCEBLE_AUTO_CONFIGURATION=yes

# Currently only available in the @dev edition see
# GH:funilrys/PyFunceble#94
export PYFUNCEBLE_OUTPUT_LOCATION="${outputDir}/"

# Export ENV variables from $HOME/.config/.pyfunceble-env

if [ -n "$useEnvPath" ]
then
    if [ -f "${condaInstallDir}/envs/${PYFUNCEBLEPACKAGENAME}/.pyfunceble-env" ]
    then
        rm "${condaInstallDir}/envs/${PYFUNCEBLEPACKAGENAME}/.pyfunceble-env"
    fi

    if [ -f "${condaInstallDir}/envs/${PYFUNCEBLEPACKAGENAME}/.PyFunceble.overwrite.yaml" ]
    then
        rm "${condaInstallDir}/envs/${PYFUNCEBLEPACKAGENAME}/.PyFunceble.overwrite.yaml"
    fi

    if [ ! -f "${condaInstallDir}/envs/${PYFUNCEBLEPACKAGENAME}/.pyfunceble-env" ]
    then
        cp "$HOME/.config/PyFunceble/.pyfunceble-env" \
            "${condaInstallDir}/envs/${PYFUNCEBLEPACKAGENAME}/.pyfunceble-env"
    fi

    if [ ! -f "${condaInstallDir}/envs/${PYFUNCEBLEPACKAGENAME}/.PyFunceble.overwrite.yaml" ]
    then
        cp "$HOME/.config/PyFunceble/.PyFunceble.overwrite.yaml" \
            "${condaInstallDir}/envs/${PYFUNCEBLEPACKAGENAME}"
    fi

    export PYFUNCEBLE_CONFIG_DIR="${condaInstallDir}/envs/${PYFUNCEBLEPACKAGENAME}/"
else
    export PYFUNCEBLE_CONFIG_DIR="${outputDir}/"
fi

# Run PyFunceble
# Switched to use array to keep quotes for SC2086
pyfunceble "${PYFUNCEBLEARGS[@]}" -f "${1}"

# Output the test variables at the end of the test, as it could have been
# Running for hours and terminal history could be to long to be visible

echo ""
echo ""
echo -e "\tThank you for feting me with all that junk food, I used to like you too..."
echo -e "\tYou tested with: $(pyfunceble --version)"
echo -e "\tYou tested this source: ${1}"
echo -e "\tWith the following variable: ${PYFUNCEBLEARGS[*]}"
echo -e "\tYour output location is: ${outputDir}"
echo -e "\tThe following files have been generated in the ${outputDir}\n"
echo ""
echo ""

tree --prune -f "${outputDir}"

# When finished - Deactivate the environment
conda deactivate

if [ "$CCLEAN" == "[yY]" ]
then
    echo "Conda.. Do you mind? Clean up after your self, please!!"
    conda env remove -n "${PYFUNCEBLEPACKAGENAME}"
fi

echo ${?}


setup-arm7l.sh
#!/usr/bin/env bash

#
# Modified by https://www.mypdns.org/p/Spirillen/
# License: https://www.mypdns.org/w/license/
# Issues: https://www.mypdns.org/maniphest/
# Project: https://www.mypdns.org/project/profile/5/
# ----------------------------------
# We will use the python 3.7 venv
# Since conda on arm7l only supports
# upto python 3.4.
# ----------------------------------

# Stop on any error
set -e

# Set python version
pythonVersion="python3.7"

installDir="${HOME}/Downloads"

branch=('4.0.0-dev' 'dev' 'master')

# First install pythonVersion if not
# already installed
# https://stackoverflow.com/questions/1298066/check-if-an-apt-get-package-is-installed-and-then-install-it-if-its-not-on-linu/54239534#54239534

pkgs="${pythonVersion} git tree"
if ! dpkg -s $pkgs >/dev/null 2>&1; then
  sudo apt-get install $pkgs
fi

if [ ! -d "${installDir}/pyfunceble" ]
then
    cd "${installDir}"
    git clone https://github.com/funilrys/PyFunceble.git pyfunceble

elif [ -d "${installDir}/pyfunceble" ]
then
    cd "${installDir}/pyfunceble"
fi

if [ -d "${installDir}/pyfunceble/venv" ]
then
    echo 'We only need to update the repo to fetch latest'
    git checkout master && git fetch
else
    for i in "${branch[@]}"
    do 
        cd "${installDir}/pyfunceble"
        echo 'We are setting up the venv'
        git checkout "${i}"
        "${pythonVersion}" -m venv venv
        source venv/bin/activate --upgrade --upgrade-deps
        pip3 install -e .
        export PYFUNCEBLE_AUTO_CONFIGURATION=yes
        pyfunceble --version
        deactivate
    done
fi

git checkout master

echo "Script exited with code: ${?}"


setup_miniconda.sh
#!/usr/bin/env bash

# All initial credits to:
# - https://github.com/mitchellkrogza
# - https://mypdns.org/funilrys
#
# Modified by https://mypdns.org/spirillen
# License: https://mypdns.org/mypdns/support/-/wikis/License
# Issues: https://mypdns.org/pyfunceble/pyfunceble-templates/pyfunceble-miniconda/-/issues
# Project: https://mypdns.org/pyfunceble/pyfunceble-templates/pyfunceble-miniconda
# -------------------------------
# Setup Conda Python Environments
# -------------------------------

# Stop on any error
set -e

# Set python version
# to find the latest version
# `conda search --full-name python`
# pythonVersion="3.10"

# Set conda install dir
condaInstallDir="${HOME}/miniconda"

function version () {
  echo ""
  echo "Python version"
  python -VVV
  
  echo ""
  echo "pip version"
  pip --version

  echo ""
  echo "PyFunceble version"
  pyfunceble --version

  # echo ""
  # echo "uhb-whitelist version"
  # uhb-whitelist --version
}

function copy_config () {
  # Copy default .pyfunceble-env and .PyFunceble.overwrite.yaml to new
  # environment
  cp "$HOME/.config/PyFunceble/.pyfunceble-env" \
    "${condaInstallDir}/envs/${PYFUNCEBLEPACKAGENAME}"

  cp "$HOME/.config/PyFunceble/.PyFunceble.overwrite.yaml" \
    "${condaInstallDir}/envs/${PYFUNCEBLEPACKAGENAME}"
  
  echo ""
  echo "List copied files"
  ls -lha "${condaInstallDir}/envs/${PYFUNCEBLEPACKAGENAME}"
}

if [[ ! -d ${condaInstallDir} ]]
then
  if [ $(uname -m) == 'x86_64' ]
  then
  # Download Conda
  wget -O miniconda.sh \
    'https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh'
  
  elif [ $(uname -m) == 'x86' ]
    then
    wget -O miniconda.sh \
    'https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86.sh'

  elif [ $(uname -m) == 'ppc64le' ]
    then
    wget -O miniconda.sh \
    'https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-ppc64le.sh'
  fi

  # Install Conda
  bash miniconda.sh -b -f -p "${condaInstallDir}"

  source "$HOME/miniconda/etc/profile.d/conda.sh"
  hash -r
  conda config --set always_yes yes --set changeps1 no
  conda update -q conda
  conda info -a
else
  echo "We assume that conda is already installed."
  echo "We are updating the environment"
  source "$HOME/miniconda/etc/profile.d/conda.sh"
  hash -r
  conda config --set always_yes yes --set changeps1 no
  conda update -q conda
  conda info -a

  echo ${?}

  echo "Updated completed"
fi

if [[ ! -d ${condaInstallDir}/envs/pyfunceble ]]
then
  # Setup pyfunceble@ version to be used
  export PYFUNCEBLEPACKAGENAME="pyfunceble"

  # Get the conda CLI.
  source "${condaInstallDir}/etc/profile.d/conda.sh"

  hash conda

  # Update Conda
  conda update -yq conda
  conda update -n base conda

  # Create an Environment (EXAMPLE: creating an environment called
  # pyfuncebletesting with Python version "pythonVersion")
  # conda create -y -q -n "${PYFUNCEBLEPACKAGENAME}" python="${pythonVersion}"
  conda env update -f ".environment.${PYFUNCEBLEPACKAGENAME}.yaml" --prune -q

  # Activate this environment you just created
  # According to the https://docs.conda.io/projects/conda/en/latest/_downloads/843d9e0198f2a193a3484886fa28163c/conda-cheatsheet.pdf
  # We shall replace source with conda activate vs source
  conda activate "${PYFUNCEBLEPACKAGENAME}"

  # Calling function: version
  version

  # uhb-whitelist --version

  # prepared for installing through conda package
  #conda install -c pyfunceble "${PYFUNCEBLEPACKAGENAME}"

  # Copy default .pyfunceble-env to new environment
  # Calling function: copy_config
  copy_config

  # When finished - Deactivate the environment
  conda deactivate

else
  echo "We assume that Pyfunceble envs is already installed."
fi

if [[ ! -d ${condaInstallDir}/envs/pyfunceble-dev ]]
then
  # Setup pyfunceble@ version to be used
  export PYFUNCEBLEPACKAGENAME="pyfunceble-dev"

  # Get the conda CLI.
  source "${condaInstallDir}/etc/profile.d/conda.sh"

  hash conda

  # Update Conda
  conda update -yq conda

  # Create an Environment (EXAMPLE: creating an environment called
  # conda create -y -q -n "${PYFUNCEBLEPACKAGENAME}" python="${pythonVersion}"
  conda env update -f ".environment.${PYFUNCEBLEPACKAGENAME}.yaml" --prune -q

  # Activate this environment you just created
  # According to the https://docs.conda.io/projects/conda/en/latest/_downloads/843d9e0198f2a193a3484886fa28163c/conda-cheatsheet.pdf
  # We shall replace source with conda activate vs source
  conda activate "${PYFUNCEBLEPACKAGENAME}"

  # Calling function: version
  version

  # Calling function: copy_config
  copy_config

  # When finished - Deactivate the environment
  conda deactivate

else
  echo "We assume that Pyfunceble-dev envs is already installed."
fi

echo "Script exited with code: ${?}"

# Now you should be ready to run the pyfunceble.sh

# For testing only
# conda env list

# conda env remove -n pyfunceble
# conda env remove -n pyfunceble-dev

# Cleanup
rm miniconda.sh


pyfunceble4-arm7.sh
#!/usr/bin/env bash
# copyeight by @spirillen
# License: https://www.mypdns.org/w/license/
# Issues: https://www.mypdns.org/maniphest/
# Project: https://www.mypdns.org/project/profile/5/

# Stop on any error
set -e #x


# Set python version
pythonVersion="python3.7"

pkgs="${pythonVersion} git tree"
if ! dpkg -s $pkgs >/dev/null 2>&1; then
  sudo apt-get install $pkgs
fi

pyfuncebleDir="${HOME}/Downloads/pyfunceble"

cd "${pyfuncebleDir}"

git checkout master
git pull

# Run this script by appending test-file to the script name in the shell prompt
# E.g. pyfunceble4-arm7.sh "/full/path/to/file"

# just a bit of fun
echo -e "\tI'm your hungry cheese missing a piece..."
echo -e "\tAre you ready to go hunting the red pills"
echo -e "\tso we can hunt down evil ghosts"
echo ""
echo -e "\tAre you ready to start?"
echo ""
echo -e "\tloading kacman..."

if [[ -z "${1}" ]]
then
    printf "\n\tYou have been eating a blue pill..."
    printf "\tThe ghosts caught you :skull:"
    printf "\tPlease show me the route to the ghosts"
    printf "\tYou want me to chew through\n\t.%s /blue/pills/is/dead/ghosts\n\n" "${0}"
    exit 1
fi

# Change the output directory to suite your needs
read -erp "Enter output directory for test results: " \
    -i "/tmp/pyfunceble4/$(date +'%H%M')" outputDir

# Clean output dir if exist for a clean test environment
if [[ -d "${outputDir}" ]]
then
    rm -fr "${outputDir}"
fi

# Set your desired pyfunceble verion
# We set pyfunceble-dev as default to avoid typos.
echo ''
echo 'What version would you like to use?'
echo '1. master'
echo '2. dev'
echo '3. 4.alpha'
echo ''

while true
do
read -erp "Which version of PyFunceble would you like to use?: pyfunceble4: " \
    -i "3" pyfuncebleVersion
case $pyfuncebleVersion in
    1)
    git checkout master
    source venv/bin/activate --upgrade --upgrade-deps
    break;; 

    2)
    git checkout dev
    source venv/bin/activate --upgrade --upgrade-deps
    break;;

    3)
    git checkout 4.0.0-dev
    source venv/bin/activate --upgrade --upgrade-deps
    break;;

    *)
    echo 'Invalid choice...'
esac
done

# Set your test string.
# IMPORTANT: the -f argument is preset as last argument

# Bug #3 test string
read -erp "Enter any custom test string: " \
    -i "-dbr 6 -ex --dns 192.168.1.104:53 -w $(nproc --ignore=2) -a --database-type mariadb --no-files --wildcard" -a pyfuncebleArgs

while true
do
read -erp "Would you like to use your default pyfunceble enviroment?: [Y/n] " -i "Y" pyfuncebleENV

case $pyfuncebleENV in
    [yY][eE][sS]|[yY])
 useEnvPath="yes"
 break
 ;;
    [nN][oO]|[nN])
 useEnvPath=""
 break
    ;;
    *)
 echo "Invalid input..."
 ;;
 esac
done

# Make sure output dir is present
mkdir -p "${outputDir}"

# print pyfunceble version
pyfunceble --version

# Tell the script to install/update the configuration file automatically.
export PYFUNCEBLE_AUTO_CONFIGURATION=yes

# Currently only availeble in the @dev edition see
# GH:funilrys/PyFunceble#94
export PYFUNCEBLE_OUTPUT_LOCATION="${outputDir}/"

# Export ENV variables from $HOME/.config/.pyfunceble-env

if [ -n "$useEnvPath" ]
then
    export PYFUNCEBLE_CONFIG_DIR="$HOME/.config/PyFunceble/.pyfunceble-env"
else
    export PYFUNCEBLE_CONFIG_DIR="${outputDir}/"
fi

# Run PyFunceble
# Switched to use array to keep quotes for SC2086
pyfunceble "${pyfuncebleArgs[@]}" -f "${1}"

# Output the test variables at the end of the test, as it could have been
# Running for hours and terminal history could be to long to be visible

echo ""
echo ""
echo -e "\tThank you for feting me with all that junk food, I used to like you too..."
echo -e "\tYou tested with: " $(pyfunceble --version)
echo -e "\tYou tested this source: ${1}"
echo -e "\tWith the following variable: ${pyfuncebleArgs[@]}"
echo -e "\tYou're output location is: ${outputDir}"
echo -e "\tThe following files have been generated in the outputDir\n"
echo ""
echo ""

tree --prune -f "${outputDir}"

echo ${?}


setup_miniconda-arm7l.sh
#!/usr/bin/env bash

#
# Modified by https://www.mypdns.org/p/Spirillen/
# License: https://www.mypdns.org/w/license/
# Issues: https://www.mypdns.org/maniphest/
# Project: https://www.mypdns.org/project/profile/5/
# ----------------------------------
# We will use the python 3.7 venv
# Since conda on arm7l only supports
# upto python 3.4.
# ----------------------------------

# Stop on any error
set -e

# Set python version
# to find the latest version
# `conda search --full-name python`
pythonVersion="3.7"

hash "${pythonVersion}" 2>/dev/null || { echo >&2 "We are installing python 3.7 for you" && sudo apt install python3.7 -y; }

cd "${HOME}/Downloads/"

git clone https://github.com/funilrys/PyFunceble.git pyfunceble

cd "${HOME}/Downloads/pyfunceble"

branch=('4.0.0-dev' 'dev' 'master')

if [ ! -d "${HOME}/Downloads/pyfunceble/venv" ]
then
    for i in "${branch[@]}"
    do 
        git checkout "${i}"
        python3.7 -m venv venv
        source venv/bin/activate --upgrade --upgrade-deps
        pip3 install -e .
        export PYFUNCEBLE_AUTO_CONFIGURATION=yes
        pyfunceble --version
        deactivate
    done
else
    git checkout master && git fetch
fi

git checkout master

pyfunceble4.sh
#!/usr/bin/env bash
# All initial credits to @mitchellkrogza and @funilrys
# Modified by @spirillen
# License: https://www.mypdns.org/w/license/
# Issues: https://www.mypdns.org/maniphest/
# Project: https://www.mypdns.org/project/profile/5/
# -------------------------------
# Setup Conda Python Environments
# -------------------------------

# Stop on any error
set -e #-x

pkgs="tree"
if ! dpkg -s $pkgs >/dev/null 2>&1; then
  sudo apt-get install $pkgs
fi

# Run this script by appending test-file to the script name in the shell prompt
# E.g. miniconda_pyfunceble.sh "/full/path/to/file"

# just a bit of fun
echo -e "\tI'm your hungry cheese missing a piece..."
echo -e "\tAre you ready to go hunting the red pills"
echo -e "\tso we can hunt down evil ghosts"
echo ""
echo -e "\tAre you ready to start?"
echo ""
echo -e "\tloading kacman..."

# Set conda install dir
condaInstallDir="${HOME}/miniconda"

if [[ -z "${1}" ]]
then
    printf "\n\tYou have been swaloowing a blue pill..."
    printf "\tThe ghosts caught you :skull:"
    printf "\tPlease show me the route to the ghosts"
    printf "\tYou want me to chew through\n\t.%s /blue/pills/is/dead/ghosts\n\n" "${0}"
    exit 1
fi

# Select Python version for test env
read -erp "Enter Python version to be used in Conda: " \
    -i "3.9.6" PV

# Change the output directory to suite your needs
read -erp "Enter output directory for test results: " \
    -i "/tmp/pyfunceble/$(date +'%H%M')" outputDir

# Clean output dir if exist for a clean test environment
if [[ -d "${outputDir}" ]]
then
    rm -fr "${outputDir}"
fi

# Set your desired pyfunceble verion
# We set pyfunceble-dev as default to avoid typos.
read -erp "Which version of PyFunceble would you like to use?: pyfunceble4: " \
    -i "pyfunceble4" PYFUNCEBLEPACKAGENAME

# Set your test string.
# IMPORTANT: the -f argument is preset as last argument

# Bug #3 test string
read -erp "Enter any custom test string: " \
    -i "--hierarchical -ex -a --dns 192.168.0.100:5302 -w 40 --database-type mariadb -dbr 0 --whois-lookup --http --netinfo-lookup" -a PYFUNCEBLEARGS

# We should change the default ENV dir to match the PyF versions conda dir
# shellcheck disable=SC2034  # Unused variables left for readability

while true
do
read -erp "Would you like to use your default pyfunceble enviroment
  ${condaInstallDir}/envs/${PYFUNCEBLEPACKAGENAME}?: [Y/n] " -i "Y" PYFUNCEBLE_ENV

case $PYFUNCEBLE_ENV in
    [yY][eE][sS]|[yY])
 useEnvPath="yes"
 break
 ;;
    [nN][oO]|[nN])
 useEnvPath=""
 break
    ;;
    *)
 echo "Invalid input..."
 ;;
 esac
done

# Get the conda CLI.
source "${condaInstallDir}/etc/profile.d/conda.sh"

hash conda

# First Update Conda
conda update -q conda

# Activate your environment
# According to the https://docs.conda.io/projects/conda/en/latest/_downloads/843d9e0198f2a193a3484886fa28163c/conda-cheatsheet.pdf
# We shall replace source with conda activate vs source
conda env update -f '.environment.yaml'

conda activate "${PYFUNCEBLEPACKAGENAME}"

# Set python version to use
conda install python="${PV}"

# Make sure output dir is present
mkdir -p "${outputDir}"

# Upgrade the environment
# pip install -I -U -q pip wheel Pygments>=2.0
# pip uninstall -yq PyFunceble-dev #"${PYFUNCEBLEPACKAGENAME}"
# pip install --no-cache-dir --upgrade -q --pre pyfunceble-dev>=4.0.0b61 'alabaster<0.8,>=0.7'
# pip install --no-cache-dir --upgrade -I -q 'git+https://github.com/funilrys/PyFunceble.git@dev'
# pip install --no-cache-dir --upgrade -I -q 'git+https://mypdns.org/pyfunceble/PyFunceble.git@dev'

if [ "${PYFUNCEBLEPACKAGENAME}" == 'pyfunceble' ]
then
	pip install --no-cache-dir --upgrade -I -q 'git+https://github.com/Ultimate-Hosts-Blacklist/whitelist.git@script'
	pyfunceble --version
	uhb-whitelist --version
	#pip list
else
	pip install --no-cache-dir --upgrade -I -q 'git+https://github.com/Ultimate-Hosts-Blacklist/whitelist.git@script-dev'
	pyfunceble --version
	uhb-whitelist --version
	#pip list
fi

# print pyfunceble version
pyfunceble --version

# Tell the script to install/update the configuration file automatically.
export PYFUNCEBLE_AUTO_CONFIGURATION=yes

# Currently only availeble in the @dev edition see
# GH:funilrys/PyFunceble#94
export PYFUNCEBLE_OUTPUT_LOCATION="${outputDir}/"

# Export ENV variables from $HOME/.config/.pyfunceble-env

if [ -n "$useEnvPath" ]
then
    if [ -f "${condaInstallDir}/envs/${PYFUNCEBLEPACKAGENAME}/.pyfunceble-env" ]
    then
        rm "${condaInstallDir}/envs/${PYFUNCEBLEPACKAGENAME}/.pyfunceble-env"
    fi

    if [ ! -f "${condaInstallDir}/envs/${PYFUNCEBLEPACKAGENAME}/.pyfunceble-env" ]
    then
        cp "$HOME/.config/PyFunceble/.pyfunceble-env.4" "${condaInstallDir}/envs/${PYFUNCEBLEPACKAGENAME}/.pyfunceble-env"
    fi
    export PYFUNCEBLE_CONFIG_DIR="${condaInstallDir}/envs/${PYFUNCEBLEPACKAGENAME}/"
else
    export PYFUNCEBLE_CONFIG_DIR="${outputDir}/"
fi

# Run PyFunceble
# Switched to use array to keep quotes for SC2086
pyfunceble "${PYFUNCEBLEARGS[@]}" -f "${1}"

# Output the test variables at the end of the test, as it could have been
# Running for hours and terminal history could be to long to be visible

echo ""
echo ""
echo -e "\tThank you for feting me with all that junk food, I used to like you too..."
echo -e "\tYou tested with: " $(pyfunceble --version)
echo -e "\tYou tested this source: ${1}"
echo -e "\tWith the following variable: ${PYFUNCEBLEARGS[@]}"
echo -e "\tYou're output location is: ${outputDir}"
echo -e "\tThe following files have been generated in the outputDir\n"
echo ""
echo ""

tree --prune -f "${outputDir}"

# When finished - Deactivate the environment
conda deactivate

echo ${?}
